src/pal/src/loader/module.cpp

```diff
@@ -1428,6 +1428,16 @@ static NATIVE_LIBRARY_HANDLE LOADLoadLibraryDirect(LPCSTR libraryNameOrPath)
     _ASSERTE(libraryNameOrPath != nullptr);
     _ASSERTE(libraryNameOrPath[0] != '\0');
 
+    size_t i = 0;
+    if (strchr(libraryNameOrPath, '/') != NULL)
+    {
+	    struct stat s;
+	    if (stat(libraryNameOrPath, &s) == -1)
```

Comment:
Have you tried using `access(2)` instead?  `stat(2)` has to validate the pointer, obtain file information, and then copy it to userland -- `access(2)` should just check if the file exists at the directory level.

However, I'm sceptical about this pattern; in fact, it's an anti-pattern, introducing a race condition known as [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use).

In fact, if you look at a `strace(1)`, especially when the system dynamic loader is linking shared libraries, you won't see this pattern. I'm not too convinced that this is in fact an optimization; and even if it is, the best place to fix it would be in `dlopen(3)` instead, benefiting the whole ecosystem.

Isn't the problem elsewhere? For instance, shouldn't `LOADLoadLibraryDirect()` be called when we're already sure the library exists to avoid `open(2)` returning an error?

Reply:
We have not tried access().

We think there is no problem with that race, is there? If another thread added a library to the disk between, say, stat and dlopen, we will not load that library, but there is no guarantee we would load it anyway.

> For instance, shouldn't LOADLoadLibraryDirect() be called when we're already sure the library exists to avoid open(2) returning an error?
Yes, we propose this optimization for throw away dozen of failing load library attempts within this cycle:
https://github.com/dotnet/coreclr/blob/master/src/vm/dllimport.cpp#L6361
but have not found any PAL abstractions for 'stat'. Can you propose one?

Reply:
`access()` looks is better choice.
Situation when one thread (or process) tries to open library and another one creates it is potentially has race and should be solved on user level. And it has race in regular `dlopen()`

Reply:
There's a `pal::file_exists()` in the hosting layer. Something similar in the runtime, calling `access(2)`, would be a good implementation.